### PR TITLE
feat(wasm): support bundled filters

### PR DESCRIPTION
**Pull request by flrgh**
### Summary

An upcoming Kong release will ship with at least one bundled Wasm filter.

Previously there was no concept of "bundled" filters: all filters must be provided by the user in the `wasm_filters_path` directory.

This introduces a new `wasm_filters` kong.conf property. It is a comma-delimited array that functions quite similarly to the `plugins` property with the addition of a special `user` value:

```
# all filters disabled
wasm_filters = off

# all bundled filters enabled
wasm_filters = bundled

# all user filters (in `wasm_filters_path`) enabled
wasm_filters = user

# all available filters enabled
wasm_filters = bundled, user

# specific filters enabled
wasm_filters = some_bundled_filter, some_user_supplied_filter
```

In the case of a collision between bundled and user filters, the conf loader will prefer the user filter and log a warning.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

KAG-4211

### CODE REVIEW

**Comment by flrgh** [comment]
The code for this is AFAICT feature-complete and should be considered ready-for-review, but this should _not_ be merged until after the distribution work for bundled filters is done.

File: kong/templates/kong_defaults.lua
Comment (hishamhm) [inline kong/templates/kong_defaults.lua:211]: can we make `bundled,user` be the default? since `wasm` is `off` by default that's harmless, and it would mean it would only take one setting (`wasm = on`) to be able to use Wasm with the bundled filter, and the setup instructions from previous Kong versions would still work for enabling user filters ("set `wasm = on` and `wasm_filters_path`")

File: kong/constants.lua
Comment (hishamhm) [inline kong/constants.lua:282]: bikeshedding: err, `/include/` feels weird as that screams "C headers". Can we use `kong/lib/filters` if we have to use a Unixy dirname? (Or straight `/usr/local/kong/filters` if we don't)

File: kong/conf_loader/init.lua
Comment (hishamhm) [inline kong/conf_loader/init.lua:650]: This is quite clunky UX, given that users would most likely not override wasm filters by accident — if such a conflict happens, they would probably be doing it on purpose with an updated version or the like. Making them reconfigure the _other_ filters because of that feels odd.

Given that `wasm_filters_path` needs to be explicitly set, I would expect user entries to take precedence over bundled entries. We could simply log an info/notice line reporting that "bundled filter x is being overriden by the one in /this/local/path/x.wasm".

File: kong/constants.lua
Comment (flrgh) [reply kong/constants.lua:282]: :+1: Yeah this is sort of a scratch value that was chosen because it resembled some known paths. We'll pick a more appropriate directory before merge.

File: kong/conf_loader/init.lua
Comment (flrgh) [reply kong/conf_loader/init.lua:650]: Felt it best to start out hyper-defensive, but I'd be okay with just logging and moving on.

File: kong/templates/kong_defaults.lua
Comment (flrgh) [reply kong/templates/kong_defaults.lua:211]: WFM, for some reason I thought `plugins` was `off` by default and wanted to match it, but I just checked and see that it is also `bundled`.

File: kong/constants.lua
Comment (flrgh) [reply kong/constants.lua:282]: ^ cc @curiositycasualty

**Comment by flrgh** [comment]
Currently rebased onto #12853. Once that is merged this PR will get a bit smaller.

File: kong/conf_loader/init.lua
Comment (hishamhm) [reply kong/conf_loader/init.lua:640]: Revisiting this: if I have bundled filters `A` and `B` and user filters `A` and `B`, there's no way to specify that I want to use bundled-`A` and user-`B` without deleting files (in either the original implemented version or my suggestion above).

In a scenario such as a custom-built container image with overriden filters, let's say someone wants to revert to bundled-`A` via configuration only without editing the image. It's a corner case but I don't think it's super far-fetched — let's say I'm trying out a custom build of a filter and I want to easily switch back and forth, but I can't disable the other user filters.

Would it be problematic to support _full paths to filters_ in `wasm_filters`, which would then take precedence over `bundled` or `user`? This would allow for:

* selecting explicitly which version of a filter we want to run
* having user wasm filters in different folders in case one needs to get them mounted via different volumes, etc.
* not having to set two variables to enable a user filter

I'm thinking basically checking whether the entry starts with `/`, and then treating that as an absolute path.

```
wasm_filters=bundled,user,/custom_filters/myfilter.wasm
```

With this, `wasm_filters_path` would essentially become a convenience in case you have multiple filters, but still probably the preferred way to use it, since it's easier to tell users "mount /my_filters/ in your container and then set `wasm_filters_path=/my_filters`.

One can't easily pull off something like this with Lua plugins because of Lua plugins reasons (Lua paths, rock dependencies, etc), but we don't need to constrain ourselves to the limitations of the `plugins` directive.

Comment (flrgh) [reply kong/conf_loader/init.lua:640]: I think if we were to support full paths in `wasm_filters`, we should just ditch `wasm_filters_path` (and the `user` keyword), and make user filters entirely path-based:

### before
```
wasm_filters = bundled,user
wasm_filters_path = /path/to/my/filters
```

### after

```
wasm_filters = bundled,/path/to/my/filters,/path/to/other/filter.wasm
wasm_filters_path = <unset>
```

Honestly I kinda like the simplicity of this. I could see past-us choosing this if we had bundled filters from the get-go.

---

This doesn't directly solve the filter name collision issue. For that, we need to decide some resolution order and decide whether collisions are an error or just a warning... OR we go the yet-another-DSL route:

I like the idea of renaming/aliasing filters in-place with `<name>=<path>`:

```
# where there is already a bundled filter named `dupe_filter`
wasm_filters = bundled,renamed_filter=/path/to/dupe_filter.wasm
```

If you just want to disable the bundled filter without renaming your own, maybe we give you a means of disabling individual filters from the bundled namespace with `!bundled:<name>`:

```
wasm_filters = bundled,!bundled:dupe_filter,/path/to/dupe_filter.wasm
```

Comment (hishamhm) [reply kong/conf_loader/init.lua:640]: As for collisions, I think it's fair to assume that any explicitly given entry overrides the version in `bundled` (as in "more specific overrides more general").

I like the idea of `wasm_filters` accepting both files and directories. I slightly dislike always having to enter `bundled,/my/path` when adding a path.

I know that's sort of how the plugins directive already works, but anything that has "it's own syntax" adds a bit of friction ("was it comma separated? does the order matter?" etc etc... the open questions accumulate; the path of the mini-DSL is the ultimate in that). In that regard, having one directive that's just a folder is always easier to remember and to get right as an end-user, so I'm fond of keeping `wasm_filters_path` around, even if it becomes just a shorthand for adding that path to the `wasm_filters` list.

File: kong/constants.lua
Comment (hishamhm) [inline kong/constants.lua:282]: should this still be here? the updated path is at `kong.conf_loader.constants`

File: spec/02-integration/20-wasm/06-clustering_spec.lua
Comment (chobits) [reply spec/02-integration/20-wasm/06-clustering_spec.lua:88]: I guess its the debugging log, see this: https://github.com/Kong/kong/pull/12843/commits/94ac12e6c026c5e22839010bbee5aee1531c0752

File: kong/conf_loader/init.lua
Comment (chobits) [reply kong/conf_loader/init.lua:641]: > As for collisions, I think it's fair to assume that any explicitly given entry overrides the version in bundled (as in "more specific overrides more general").

Additionally, you can give the user a warning, like `[warn] ... The user plugin <XXX> has overridden the bundled one`, and allow the execution to proceed.

> If you just want to disable the bundled filter without renaming your own, maybe we give you a means of disabling individual filters from the bundled namespace with !bundled:<name>:
>```
>wasm_filters = bundled,!bundled:dupe_filter,/path/to/dupe_filter.wasm
>```

This seems to add a bit of complexity to the use and implementation of  the`wasm_filters` option.

**Comment by kikito** [comment]
This pr has some debug messages on it, I am moving it to Draft until that is addressed. @flrgh please move it back to "Ready to review" when ready.

**Comment by brentos** [comment]
I'm a bit late for adding my $0.02 on this, but I think some sort of namespacing of filters could nicely solve the collision issue. For example:
```
kong.datakit
kong.response_transformer
foocompany.datakit
foocompany.response_transfomer
```
Then we could enable with 
```
wasm_filters = kong.datakit, foocompany.datakit
```
In addition, setting up some standard file locations for these such that we don't have to configure a base path, and these filters could be installed at:
```
<default_filter_dir>/kong/datakit
<default_filter_dir>/foocompany/datakit
```

File: spec/02-integration/20-wasm/06-clustering_spec.lua
Comment (flrgh) [inline spec/02-integration/20-wasm/06-clustering_spec.lua:180]: re: all of these timeout changes

This is to account for the increased NGINX startup time incurred from loading the new `datakit` module. Locally I'm seeing about ~3.6s pretty consistently, but timing in CI (where we are often saturating the CPU) is slower and very inconsistent, so a large timeout is selected to keep the test from being flaky.

There is some work in progress to mitigate the slow startup time: https://github.com/Kong/ngx_wasm_module/pull/536

File: changelog/unreleased/kong/wasm-bundled-filters.yml
Comment (brentos) [inline changelog/unreleased/kong/wasm-bundled-filters.yml:1]:
```suggestion
message: Add `wasm_filters` configuration variable for enabling individual filters
```

**Review by brentos — APPROVED** [review]

**Comment by flrgh** [comment]
investigating something, please do not merge

File: spec/02-integration/20-wasm/06-clustering_spec.lua
Comment (flrgh) [reply spec/02-integration/20-wasm/06-clustering_spec.lua]: update: because module loading is an init_worker thing, the privileged agent worker is _also_ delayed on startup, so this affects how long it takes for the DP to establish a connection to the CP. I believe that is why this test suite is struggling so much more than others. Changes:

1. added some more logfile assertions to watch for a module load event (technically this only guarantees that 1/N processes has finished init_worker, but it's still somewhat helpful).
2. adjusted the worker procs to 2
3. disabled bundled filters in this test suite because `datakit` is the biggest source of slow init_worker

**Review by hishamhm — APPROVED** [review]

**Pull request merged**
